crates/services/txpool_v2/src/storage/graph.rs

```diff
@@ -0,0 +1,440 @@
+                    if to != i_owner {
+                        return Err(Error::NotInsertedIoWrongOwner);
+                    }
+                    if amount != i_amount {
+                        return Err(Error::NotInsertedIoWrongAmount);
+                    }
+                    if asset_id != i_asset_id {
+                        return Err(Error::NotInsertedIoWrongAssetId);
+                    }
+                }
+                Output::Contract(_) => return Err(Error::NotInsertedIoContractOutput),
+                Output::Change { .. } => {
+                    return Err(Error::NotInsertedInputDependentOnChangeOrVariable)
+                }
+                Output::Variable { .. } => {
+                    return Err(Error::NotInsertedInputDependentOnChangeOrVariable)
+                }
+                Output::ContractCreated { .. } => {
+                    return Err(Error::NotInsertedIoContractOutput)
+                }
+            };
+        }
+        Ok(())
+    }
+
+    /// Cache the transaction information in the storage caches.
+    /// This is used to speed up the verification/dependencies searches of the transactions.
+    fn cache_tx_infos(
+        &mut self,
+        outputs: &[Output],
+        tx_id: &TxId,
+        node_id: NodeIndex,
+    ) -> Result<(), Error> {
+        for (index, output) in outputs.iter().enumerate() {
+            let index = u16::try_from(index).map_err(|_| {
+                Error::WrongOutputNumber(format!(
+                    "The number of outputs in `{}` is more than `u8::max`",
+                    tx_id
+                ))
+            })?;
+            let utxo_id = UtxoId::new(*tx_id, index);
+            match output {
+                Output::Coin { .. } | Output::Change { .. } | Output::Variable { .. } => {
+                    self.coins_creators.insert(utxo_id, node_id);
+                }
+                Output::ContractCreated { contract_id, .. } => {
+                    self.contracts_creators.insert(*contract_id, node_id);
+                }
+                _ => {}
+            }
+        }
+        Ok(())
+    }
+
+    /// Clear the caches of the storage when a transaction is removed.
+    fn clear_cache(&mut self, outputs: &[Output], tx_id: &TxId) -> Result<(), Error> {
+        for (index, output) in outputs.iter().enumerate() {
+            let index = u16::try_from(index).map_err(|_| {
+                Error::WrongOutputNumber(format!(
+                    "The number of outputs in `{}` is more than `u8::max`",
+                    tx_id
+                ))
+            })?;
+            let utxo_id = UtxoId::new(*tx_id, index);
+            match output {
+                Output::Coin { .. } | Output::Change { .. } | Output::Variable { .. } => {
+                    self.coins_creators.remove(&utxo_id);
+                }
+                Output::ContractCreated { contract_id, .. } => {
+                    self.contracts_creators.remove(contract_id);
```

Comment:
Caches are not critical so I don't think we should error in case there is nothing in cache. However, it might be nice to improve the safety of the syncronization between value and cache (cf. PR description)